WritingTools tests are failing/crashing on mac debug builds.

megangardner opened the pull request:
#### 19ad276d70970698f0dcfcf1bbec8d65ff8b59af
<pre>
WritingTools tests are failing/crashing on mac debug builds.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276405">https://bugs.webkit.org/show_bug.cgi?id=276405</a>
<a href="https://rdar.apple.com/131429390">rdar://131429390</a>

Reviewed by Richard Robinson.

We need to wait a bit longer than a presentation update to let the animations
run long enough for the test to be replaced, so that the next call is
operating on the correct ranges.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingWithStreamingSuggestions)):
(TEST(WritingTools, ProofreadingWithAttemptedEditing)):
(TEST(WritingTools, CompositionWithAttemptedEditing)):
(TEST(WritingTools, CompositionWithUndoAfterEnding)):
(TEST(WritingTools, CompositionWithUndo)):
(TEST(WritingTools, CompositionWithMultipleUndosAndRestarts)):
(TEST(WritingTools, CompositionWithUndoAndRestart)):
(TEST(WritingTools, CompositionWithAttributedStringAttributes)):
(TEST(WritingTools, CompositionWithImageRoundTrip)):
(TEST(WritingTools, CompositionWithNonImageAttachmentRoundTrip)):
(TEST(WritingTools, CompositionWithSystemFont)):
(TEST(WritingTools, CompositionWithMultipleChunks)):
(TEST(WritingTools, CompositionWithTrailingNewlines)):
(TEST(WritingTools, CompositionWithTrailingBreaks)):
(TEST(WritingTools, SmartRepliesMatchStyle)):
(TEST(WritingTools, ContextRangeFromCaretSelection)):
(TEST(WritingTools, ContextRangeFromRangeSelection)):
(TEST(WritingTools, SuggestedTextIsSelectedAfterSmartReply)):

Canonical link: <a href="https://commits.webkit.org/280870@main">https://commits.webkit.org/280870@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86eb4081efccc995aeed75845fc66fa246d54e61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37027 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61321 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8144 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59827 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44663 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8332 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46748 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5771 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34721 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49856 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27573 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7147 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7148 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63002 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1613 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7495 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53968 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1619 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54081 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12803 "") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1363 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8630 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32856 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33942 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -398,6 +398,8 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
         EXPECT_WK_STREQ(originalText, contexts.firstObject.attributedText.string);
 
         [[webView writingToolsDelegate] proofreadingSession:session.get() didReceiveSuggestions:@[ adoptNS([[WTTextSuggestion alloc] initWithOriginalRange:NSMakeRange(0, 4) replacement:@"ZZ"]).get() ] processedRange:NSMakeRange(0, 4) inContext:contexts.firstObject finished:NO];
+        TestWebKitAPI::Util::runFor(0.1_s);
```
maybe instead of a fixed timeout, there can be a new helper method like `-[WKWebView waitForUpdatedContents]` that can be used in place of this, and just loops until the contents are different. This could also replace your existing `waitForContentValue` method which would also help debugging since errors would show the mismatch instead of just timing out.

megangardner replied:
I don’t understand the second sentence. If I just waiting for content changes, that’s just timing out different if they don’t change.

rr-codes replied:
ya if they don't change at all, they'll still time out. But if they change to something that isn't the expected results (which is a relatively frequent type of fail with these types of tests when debugging), they will give an error message with the mismatch instead of timing out.

megangardner replied:
I still don't understand. Are you saying that waitForUpdatedContents should take what we expect it to change _to_, not just wait until the content changes to _anything else_. If so, I can see how that might throw an error. If not, then I don't see any helpful error that it could throw.

rr-codes replied:
For example, in this use case

```
        [[webView writingToolsDelegate] compositionSession:session.get() didReceiveText:longerAttributedText.get() replacementRange:NSMakeRange(0, 910) inContext:contexts.firstObject finished:YES];


        [webView waitForContentValue:result];
        EXPECT_WK_STREQ(result, [webView contentsAsStringWithoutNBSP]);
```

if the web view contents do change, but to a value other than `result`, in the current implementation, it will just wait and then time out, and make debugging a bit tricky. In my proposed solution, it will not time out, and instead go on to the next line, which will compare the two values and give an error message that they are different, along with the actual difference between them.

I can make this change in a follow-up PR though so that this can get merged and stop the failing tests.

megangardner replied:
This still won't help in the case where the content never actually changes, Which is what I think it happening in many of the cases. But I could be wrong.

rr-codes reviewed (changes requested):
Actually, I think this PR should also try to fix the `CompositionWithMultipleUndosAndRestarts` test too, since that seems to still be failing (assertion at TextAnimationController.mm(186))

megangardner committed:
WritingTools tests are failing/crashing on mac debug builds.
https://bugs.webkit.org/show_bug.cgi?id=276405
rdar://131429390

Reviewed by Richard Robinson.

We need to wait a bit longer than a presentation update to let the animations
run long enough for the test to be replaced, so that the next call is
operating on the correct ranges.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingWithStreamingSuggestions)):
(TEST(WritingTools, ProofreadingWithAttemptedEditing)):
(TEST(WritingTools, CompositionWithAttemptedEditing)):
(TEST(WritingTools, CompositionWithUndoAfterEnding)):
(TEST(WritingTools, CompositionWithUndo)):
(TEST(WritingTools, CompositionWithMultipleUndosAndRestarts)):
(TEST(WritingTools, CompositionWithUndoAndRestart)):
(TEST(WritingTools, CompositionWithAttributedStringAttributes)):
(TEST(WritingTools, CompositionWithImageRoundTrip)):
(TEST(WritingTools, CompositionWithNonImageAttachmentRoundTrip)):
(TEST(WritingTools, CompositionWithSystemFont)):
(TEST(WritingTools, CompositionWithMultipleChunks)):
(TEST(WritingTools, CompositionWithTrailingNewlines)):
(TEST(WritingTools, CompositionWithTrailingBreaks)):
(TEST(WritingTools, SmartRepliesMatchStyle)):
(TEST(WritingTools, ContextRangeFromCaretSelection)):
(TEST(WritingTools, ContextRangeFromRangeSelection)):
(TEST(WritingTools, SuggestedTextIsSelectedAfterSmartReply)):

Canonical link: https://commits.webkit.org/280870@main
